### PR TITLE
Automatically cleanup git repos on the Artemis server

### LINKED ISSUE #2286: Automatically clear old repositories from repos folder.
### Is your feature request related to a problem?
Copying the `repos` folder (e.g. when moving from one VM to another) takes very long because of all the small files within the `.git` folder.
This is quite frustrating (cc @Mtze).

### Describe the solution you'd like
Clear old repositories (e.g. when the exercise is over), we can always check the repository out from the VCS e.g. in case the students wants to practice.

### Describe alternatives you've considered
Add a button to clear all old repositories from all exercises.

### PR DESCRIPTION
**Pull request by krusche**
Fix #2286

### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I followed the [coding and design guidelines](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/server.html).
- [x] Server: I added multiple integration tests (Spring) related to the features (with a high test coverage)
- [x] Server: I added `@PreAuthorize` and check the course groups for all new REST Calls (security)
- [x] Server: I implemented the changes with a good performance and prevented too many database calls
- [x] Server: I documented the Java code using JavaDoc style.

### Motivation and Context
Simplify maintenance

### Description
Delete old git repos on the Artemis server (not on Bitbucket / Gitlab) to save space

### Steps for Testing
You cannot really test this :-(
However, code review is welcome!

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/service/scheduled/AutomaticProgrammingExerciseCleanupService.java [hunk 1]
```diff
@@ -4,47 +4,103 @@
...
 import de.tum.in.www1.artemis.repository.ProgrammingExerciseStudentParticipationRepository;
+import de.tum.in.www1.artemis.security.SecurityUtils;
 import de.tum.in.www1.artemis.service.ParticipationService;
+import de.tum.in.www1.artemis.service.connectors.GitService;
 import io.github.jhipster.config.JHipsterConstants;
 
 @Service
-@Profile("scheduling")
-public class AutomaticBuildPlanCleanupService {
+// @Profile("scheduling")
```
Comment (sleiss) [inline src/main/java/de/tum/in/www1/artemis/service/scheduled/AutomaticProgrammingExerciseCleanupService.java:30]: Is this change intended? As (at least in our setup) all instances share the filesystem, it should be enough if one node does the cleanup. With this change, the cleanup on the CI server will also be triggered by every instance, correct?

Comment (krusche) [reply src/main/java/de/tum/in/www1/artemis/service/scheduled/AutomaticProgrammingExerciseCleanupService.java:30]: good catch, this was only for testing purposes, I'll fix it

**Review by SimonEntholzer — DISMISSED** [review]
Querries look good, other code changes look fine too.

File: src/main/java/de/tum/in/www1/artemis/repository/ProgrammingExerciseRepository.java [hunk 2]
```diff
@@ -54,13 +54,19 @@
             + "and (spr.id = (select max(id) from sp.results) or spr.id = null)")
     Optional<ProgrammingExercise> findWithTemplateAndSolutionParticipationLatestResultById(@Param("exerciseId") Long exerciseId);
 
-    @Query("select distinct pe from ProgrammingExercise as pe left join fetch pe.studentParticipations")
+    @Query("select distinct pe from ProgrammingExercise pe left join fetch pe.studentParticipations")
```
Comment (FuchsDominik) [inline src/main/java/de/tum/in/www1/artemis/repository/ProgrammingExerciseRepository.java:57]:
```suggestion
    @Query("""
    select distinct pe 
    from ProgrammingExercise pe left join fetch pe.studentParticipations
    """)
```

**Review by FuchsDominik — DISMISSED** [review]
Looks good to me. One (not critical) thing I noticed is that we actually want to use the
`("""
   *QUERY*
   """)` style for our new SQL queries

**Review by derLalla — APPROVED** [review]
I didn't find any issues with the code, but I am also agreeing with @FuchsDominik that we should update the SQL queries if possible.

File: src/main/java/de/tum/in/www1/artemis/service/connectors/GitService.java [hunk 3]
```diff
@@ -918,6 +918,23 @@ public void deleteLocalRepository(VcsRepositoryUrl repoUrl) {
...
+     * delete the folder in the file system that contains all repositories for the given programming exercise
+     * @param programmingExercise contains the project key which is used as the folder name
+     */
+    public void deleteLocalProgrammingExerciseReposFolder(ProgrammingExercise programmingExercise) {
+        var folderPath = Paths.get(repoClonePath, programmingExercise.getProjectKey());
+        try {
+            FileUtils.deleteDirectory(folderPath.toFile());
+        }
+        catch (IOException ex) {
+            log.error("Exception during deleteLocalProgrammingExerciseReposFolder " + ex, ex);
```
Comment (ivanchimeno) [inline src/main/java/de/tum/in/www1/artemis/service/connectors/GitService.java:931]: Not sure if it's important but should it be `ex.getMessage()` like the changes written below?

**Review by ivanchimeno — APPROVED** [review]
The code looks good for me, I approve

**Review by FrankeLukas — DISMISSED** [review]
Code changes look good to me.

File: src/main/java/de/tum/in/www1/artemis/repository/ProgrammingExerciseRepository.java (hunk 2 above)
Comment (FrankeLukas) [reply src/main/java/de/tum/in/www1/artemis/repository/ProgrammingExerciseRepository.java:57]: Most queries in this file are one-liners. Imo we should either change all for consistency or leave this as is.

**Review by FrankeLukas — APPROVED** [review]
Code issue(s) resolved.

**Pull request merged**